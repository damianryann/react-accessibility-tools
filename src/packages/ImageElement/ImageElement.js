import React from "react";
import PropTypes from "prop-types";

const ImageElement = (props) => {
  const { src, alt, hideFromReader, className, id, style } = props;

  function addClass() {
    let classToImg = (document.getElementsByTagName(
      "img"
    ).className = className);

    return classToImg;
  }

  function addId() {
    let idToImg = (document.getElementsByTagName("img").id = id);

    return idToImg;
  }

  function addStyle() {
    let styleToImg = (document.getElementsByTagName("img").style = style);

    return styleToImg;
  }

  function addAriaHidden() {
    if (hideFromReader) {
      let ariaToImg = (document.getElementsByTagName("img").ariaHidden =
        "true");

      return ariaToImg;
    }
  }

  const imageElementMarkup = !alt ? (
    <img
      src={src}
      role="presentation"
      className={addClass()}
      id={addId()}
      style={addStyle()}
      alt=""
    />
  ) : (
    <img
      src={src}
      className={addClass()}
      id={addId()}
      style={addStyle()}
      alt={alt}
      aria-hidden={addAriaHidden()}
    />
  );

  return imageElementMarkup;
};

ImageElement.propTypes = {
  src: PropTypes.string,
  alt: PropTypes.string,
  hideFromReader: PropTypes.bool,
  className: PropTypes.string,
  id: PropTypes.string,
  style: PropTypes.object,
};

export default ImageElement;
