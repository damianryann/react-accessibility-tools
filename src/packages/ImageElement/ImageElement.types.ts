import { CSSProperties } from 'react';

export interface ImageElementProps {
  src: string;
  alt: string | '';
  className?: string;
  id?: string;
  style?: CSSProperties | undefined;
}
