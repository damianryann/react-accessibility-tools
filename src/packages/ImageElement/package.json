{
  "name": "@damianryann/image-element",
  "version": "1.0.0",
  "main": "ImageElement.tsx",
  "author": "Damian Ryan <damian@damianryan.co.uk>",
  "license": "MIT"
}