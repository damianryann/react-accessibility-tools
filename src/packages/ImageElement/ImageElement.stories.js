import React from "react";

import ImageElement from "./ImageElement";

import img from "../../assets/images/react.png";
import "../../sass/index.scss";

export default { title: "ImageElement" };

export const withAltText = () => <ImageElement src={img} alt="React Logo" />;
export const withoutAltText = () => <ImageElement src={img} alt="" />;
export const withoutAltDecortative = () => (
  <ImageElement src={img} alt="React Logo" hideFromReader={true} />
);
export const withExtras = () => (
  <ImageElement
    src={img}
    alt="React Logo"
    className="text-center"
    id="abc-123"
    style={{ backgroundColor: "#101010", width: "10%" }}
  />
);
