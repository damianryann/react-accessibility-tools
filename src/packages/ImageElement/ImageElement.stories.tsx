import React from 'react';

import img from '../../assets/logo.svg';
import '../../sass/index.scss';
import ImageElement from './ImageElement';
import { ImageElementProps } from './ImageElement.types';

export default { title: 'React Accessible Components/ImageElement', component: ImageElement };

export const withAltText = ({ src, alt, ...props }: ImageElementProps) => (
  <div className="w-50">
    <ImageElement src={src ?? img} alt={alt} {...props} />
  </div>
);

export const withoutAltDecortative = ({
  src,
  alt,
  ...props
}: ImageElementProps) => (
  <div className="w-50">
    <ImageElement src={src ?? img} alt={alt} {...props} />{' '}
  </div>
);

export const withExtras = ({ src, alt, ...props }: ImageElementProps) => (
  <div className="w-50">
    <ImageElement src={src ?? img} alt={alt} {...props} />{' '}
  </div>
);

withAltText.args = {
  src: img,
  alt: 'This is alt text'
};

withoutAltDecortative.args = {
  src: img
};

withExtras.args = {
  src: img,
  alt: 'This also has alt text',
  className: 'border border-2 border-primary',
  id: 'accessible-image-id'
};
