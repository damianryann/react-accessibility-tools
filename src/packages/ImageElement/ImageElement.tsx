import React, { FunctionComponent } from 'react';

import { ImageElementProps } from './ImageElement.types';

const ImageElement: FunctionComponent<ImageElementProps> = ({
  src,
  alt,
  className,
  id,
  style
}) => {
  const imageElementMarkup = alt ? (
    <img
      src={src}
      className={className ?? undefined}
      id={id ?? undefined}
      style={style ?? undefined}
      alt={alt}
    />
  ) : (
    <img
      src={src}
      role="presentation"
      className={className ?? undefined}
      id={id ?? undefined}
      style={style ?? undefined}
      alt=""
    />
  );

  return imageElementMarkup;
};

export default ImageElement;
