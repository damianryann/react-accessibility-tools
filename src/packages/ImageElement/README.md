### ImageElement

**Available Props:**

| Name             | Type      | Description                                                |
| ---------------- | --------- | ---------------------------------------------------------- |
| `hideFromReader` | `Boolean` | Is the image alt text decoration but needed for SEO?       |
| `src`            | `String`  | **Required.** Image source                                 |
| `alt`            | `String`  | Alt text for image                                         |
| `className`      | `String`  | Add additional classes for styling.                        |
| `id`             | `String`  | Add an ID for behaviours such as anchors.                  |
| `style`          | `Object`  | Use inline styling should you need to dynamically add CSS. |

```js
import React from "react";
import ImageElement from "react-accessibility-tools";

const MediaText = (props) => {
  const { data = {} } = props;
  const { ariaTrue, thumbnail, logicalName, altText } = data;

  return (
    <div className="container">
      <div className="row">
        <div className="col-md-12">
          <ImageElement
            src={thumbnail}
            alt={altText} //if empty, output will display role="presentation"
            id={`${logicalName}-thumbnail`}
            className="img-thumbnail"
            hideFromReader={ariaTrue} //only true if altText exists and bool toggled.
          />
        </div>
      </div>
    </div>
  );
};

export default MediaText;
```
