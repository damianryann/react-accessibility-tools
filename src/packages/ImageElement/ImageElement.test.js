import React from "react";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { mount } from "enzyme";

import ImageElement from "./ImageElement";

Enzyme.configure({ adapter: new Adapter() });

it("mounts without crashing", () => {
  const component = mount(<ImageElement />);

  expect(component).toMatchSnapshot();
});

it("expect no changes when alt text is present", () => {
  const component = mount(
    <ImageElement src="../hello.png" alt="hello world" />
  );

  expect(component.find("img")).toHaveLength(1);
  expect(component.find("img").html()).toBe(
    `<img src="../hello.png" alt="hello world">`
  );

  expect(component).toMatchSnapshot();
});

it("expect role=presentation to be active when no alt text is provided", () => {
  const component = mount(<ImageElement src="../hello.png" alt="" />);

  expect(component.find("img")).toHaveLength(1);
  expect(component.find("img").html()).toBe(
    `<img src="../hello.png" role="presentation" alt="">`
  );

  expect(component).toMatchSnapshot();
});

it("expect aria hidden to be true if alt text is for SEO only", () => {
  const component = mount(
    <ImageElement src="../hello.png" alt="hello world" hideFromReader={true} />
  );

  expect(component.find("img")).toHaveLength(1);
  expect(component.find("img").html()).toBe(
    `<img src="../hello.png" alt="hello world" aria-hidden="true">`
  );

  expect(component).toMatchSnapshot();
});

it("classname and id is added to image element", () => {
  const component = mount(
    <ImageElement className="text-center" id="abc-123" />
  );

  expect(component.find("img").hasClass("text-center")).toBeTruthy();
  expect(component.find("#abc-123")).toBeTruthy();

  expect(component).toMatchSnapshot();
});

it("style can be added to image", () => {
  const component = mount(
    <ImageElement
      src="../hello.png"
      alt="hello world"
      style={{ backgroundColor: "#ffffff" }}
    />
  );

  expect(component.find("img")).toHaveLength(1);
  expect(component.find("img").html()).toBe(
    `<img src="../hello.png" style="background-color: rgb(255, 255, 255);" alt="hello world">`
  );

  expect(component).toMatchSnapshot();
});
