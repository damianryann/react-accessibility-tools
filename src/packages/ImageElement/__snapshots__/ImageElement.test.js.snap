// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classname and id is added to image element 1`] = `ReactWrapper {}`;

exports[`expect aria hidden to be true if alt text is for SEO only 1`] = `ReactWrapper {}`;

exports[`expect no changes when alt text is present 1`] = `ReactWrapper {}`;

exports[`expect role=presentation to be active when no alt text is provided 1`] = `ReactWrapper {}`;

exports[`mounts without crashing 1`] = `ReactWrapper {}`;

exports[`style can be added to image 1`] = `ReactWrapper {}`;
